src/internal/document_api/mod.rs

```diff
@@ -435,53 +500,177 @@ pub fn encrypt_document<
-        .join4(
+        .join3(
+            // TICKET: only make these calls if the vecs are non-empty
             internal::group_api::get_group_keys(auth, group_grants),
             policy_grant.map(|p| requests::policy_get::policy_get_request(auth, p)),
-            aes::encrypt_future(rng, &plaintext.to_vec(), *doc_sym_key.bytes()),
         )
-        .and_then(move |(users, groups, maybe_policy_res, encrypted_doc)| {
+        .map(move |(users, groups, maybe_policy_res)| {
             let (group_errs, groups_with_key) = process_groups(groups);
             let (user_errs, users_with_key) = process_users(users);
-            let explicit_grants = if grant_to_author {
-                let self_grant = WithKey::new(
-                    UserOrGroup::User {
-                        id: auth.account_id.clone(),
-                    },
-                    user_master_pub_key.clone(),
-                );
-                [&[self_grant], &users_with_key[..], &groups_with_key[..]].concat()
-            } else {
-                [&users_with_key[..], &groups_with_key[..]].concat()
-            };
+            let explicit_grants = [users_with_key, groups_with_key].concat();
             let (policy_errs, applied_policy_grants) = match maybe_policy_res {
                 None => (vec![], vec![]),
                 Some(res) => process_policy(&res),
             };
+            let maybe_self_grant = {
+                move |public_key: &PublicKey| {
+                    if grant_to_author {
+                        vec![WithKey::new(
+                            UserOrGroup::User {
+                                id: auth.account_id.clone(),
+                            },
+                            public_key.clone(),
+                        )]
+                    } else {
+                        vec![]
+                    }
+                }
+            };
 
-            // squish all accumulated errors into one list
-            let other_errs = vec![group_errs, user_errs, policy_errs]
-                .into_iter()
-                .concat();
-
-            encrypt_document_core(
-                auth,
-                recrypt,
-                dek,
-                encrypted_doc,
-                doc_id,
-                document_name,
-                [explicit_grants, applied_policy_grants].concat(),
-                other_errs,
+            (
+                {
+                    move |pk: &PublicKey| {
+                        [maybe_self_grant(pk), explicit_grants, applied_policy_grants].concat()
+                    }
+                },
+                [group_errs, user_errs, policy_errs].concat(),
             )
         })
 }
 
+/// Encrypts a document but does not create the document in the IronCore system.
+/// The resultant DocumentDetachedEncryptResult both the EncryptedDeks and the AesEncryptedValue for the caller to deal with.
+pub fn edek_encrypt_document<'a, R1, R2: 'a>(
```

Comment:
seems a lot of shared code between this and `encrypt_document` - could the common code be factored out into a new function?

Reply:
Maybe? Much of the work in this branch was to refactor the duplicated encrypt code. The result of these refactorings are `encrypte_document_core` and `resolve_keys_for_grants`. After another glance I don't see any big gains to be had by further code sharing between `edek_encrypt_docuement` and `encrypt_document`. Certainly open to suggestions thought.